\chapter*{Abstract}
\addcontentsline{toc}{chapter}{Abstract}

Basic Local Alignment Search Tool (BLAST) has been widely used since 1990 to align DNA sequences.
Since the Next-Generation Sequencing (NGS) revolution, sequencing data has been mapped with faster software like Burrows-Wheeler Aligner (BWA).
In this project, we propose to see if BLAST can have a better accuracy and therefore can be used, in some applications, to complement existing aligners.

We created two programs, \fastqtofasta{} and \blastobam{}, to enable the use of NGS data by BLAST.
\fastqtofasta{} transforms the NGS format fastQ into a fasta file so that the sequences can be loaded into BLAST.
\blastobam{} extracts the alignment data from BLAST and outputs a file in the standard NGS format: Sequence Alignment\slash\hspace{0pt}Map format (SAM).
\blastobam{} is able to produce paired records if paired-end sequencing was performed.
The output has passed validation tests.

We then evaluated on diverse parameters the differences between the results of BLAST--\blastobam{} and BWA.
As expected, BWA proved to be much faster.
However, BLAST--\blastobam{} demonstrated a better accuracy by displaying an interesting increase in the coverage of the sequencing.

In conclusion, we showed that BLAST--\blastobam{} may be useful in some NGS applications to confirm and\slash\hspace{0pt}or to improve other aligners' results.
\bigskip

\textbf{Keywords:} BLAST, BWA, NGS, SAM.
